### PR TITLE
nrf: Merge hex files into firmware.hex.

### PR DESCRIPTION
**Pull request by aykevl**
This can make flashing a bit easier by only having to flash a single .hex file with both the SoftDevice and the application firmware.

This PR is split off from the DFU PR, to make things easier to review.

### CODE REVIEW

**Comment by glennrub** [comment]
Thanks Ayke! This is a good idea. 

Looking at the script supplied in the PR reminds me of the python intelhex's  supplied script; http://python-intelhex.readthedocs.io/en/latest/part3-4.html. The scripts gets installed if you install IntelHex using pip.

Could it be an alternative to just use this one? Personally i have used this myself as one-liner to merge SoftDevice and Application with the correct offset.

**Comment by aykevl** [comment]
Ah, that looks nice. It has a slightly different approach, though: that script appears to overwrite byte ranges when they clash, while my hex file merger emulates flash pages: when the second hex file has data in the same page as the first the whole page is erased (to emulate the effect of flashing multiple hex files after another). I'm not sure what the best method is.

In any case, I'm a bit cautious with adding an external dependency that has to be installed for all builds.

File: ports/nrf/Makefile [hunk 1]
```diff
@@ -242,47 +247,45 @@ $(filter $(PY_BUILD)/../extmod/vfs_fat_%.o, $(PY_O)): COPT += -Os
...
-	$(OBJCOPY) -O binary $< $@
+$(BUILD)/firmware.bin: $(BUILD)/firmware.hex
+	$(OBJCOPY) -O binary -I ihex $< $@
 
 ## Create binary .hex file from the .out file
-hex: $(BUILD)/$(OUTPUT_FILENAME).hex
+hex: $(BUILD)/firmware.hex
+
+## Join all hex files together to produce one flashable file
+$(BUILD)/firmware.hex: $(SOFTDEV_HEX) $(BUILD)/application.hex
```
Comment (glennrub) [inline ports/nrf/Makefile:260]: I wonder if this is always correct. Some devices like feather52 might have a static SOFTDEV_HEX pre-flashed. Could there be one alternative target where this merge is not enforced?

Comment (aykevl) [reply ports/nrf/Makefile:260]: Good point! Of course you could just flash application.hex - but that's not really a solution.
I'll think about this.

Comment (aykevl) [reply ports/nrf/Makefile:260]: Is this still an issue, now that feather52 doesn't use OTA anymore?

Comment (glennrub) [reply ports/nrf/Makefile:260]: should not be an issue anymore.

**Comment by glennrub** [comment]
I did some testing here, and these are the results i get: 
make SD=s132 flash -> application.hex (app only)
make SD=s132 sd -> firmware.hex (combined)

Looking at this a second time, i'm not sure what is real change here, comparing it with todays "sd" target, which also flashes both the app and SD as a single make target. Is the key here that you would like to do the mergehex as a default, and later use this to do a DFU of the SD+app?

**Comment by glennrub** [comment]
In order to clean up a bit i'm closing this PR as the continuation of the nrf-port development should happen in mainline micropython repository after the merge.

**Pull request closed without merging**